### PR TITLE
Function spaces implementation : starting with infinite dimensional spheres

### PR DESCRIPTION
**Pull request by kiranvad**
Hello,
This is a follow-up for #1194 (apologies for the extreme delay on this)
This PR:
* Introduces function spaces module with the implementation of infinite-dimensional spheres with L2 metrics that are mentioned in Chapter 4.3.2 in the book [Functional and Shape Data Analysis](https://link.springer.com/book/10.1007/978-1-4939-4020-2).
* This class will later be used to aid the implementation of general probability measures under the Fisher-Rao metric and SRSF for functions to be added later.

### CODE REVIEW

**Comment by ninamiolane** [comment]
Excellent @kiranvad , thank you very much! We can look at the code once you request our reviews. In the meantime, it seems that some GitHub actions are failing: could you have a look? Thanks!

**Comment by kiranvad** [comment]
Hi @ninamiolane, this PR should be ready for review now. Looking forward to your feedback!

**Comment by ninamiolane** [comment]
Looking great @kiranvad , thank you for this great addition ! Some "Linting and Testing" workflow are still failing though, could you have a look (click on "Details" to see what is the bug) before we get a human code reviewer on this? Many thanks! 🙌 

Don't worry about the "Docs check and Deployment" workflow though, you do not need to fix this one.

Also, very very minor but the formatting of the docstrings is not exact (just missing blank lines), see here:
https://github.com/geomstats/geomstats/blob/master/docs/contributing.rst#the-anatomy-of-a-docstring

(Respecting this exact docstring format helps generating the doc website automatically).

**Comment by kiranvad** [comment]
Hi @ninamiolane , thanks for the feedback. 
I have a very quick question regarding the testing workflows. It appears that test functions attributes from the `setup_method` are not being accessible from other methods in the file/module (that's the error and I can reproduce it on my local machine for other functions in the tests folder. I tried this with `test_euclidean`). 
On my local machine, I can fix this by renaming `setup_method` to the usual `setUp`.
Running `nose2 --verbose tests.tests_geomstats.test_functions` then works perfectly fine for me.
I am not familiar with fixing these issues so any help would be greatly appreciated.

**Comment by ninamiolane** [comment]
@kiranvad sorry, this is our bad! Our contributing guidelines were still indicating to use `nose2` as our testing suite, whereas we actually moved to `pytest` now. I am updating the contributing guidelines in this PR #1262 . So sorry, I hope you did not loose too much time on this. 

`setUp` was used by `nose2`,  `setup_method` is used by `pytest`. 

Could you try again with `pytest` and `setup_method`? I didn't see anything wrong with your code otherwise.

File: tests/tests_geomstats/test_functions.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import numpy as np
+
+import geomstats.backend as gs
+import geomstats.tests
+from geomstats.geometry.functions import SinfSpace
+
+
+class TestFunctionsSinf(geomstats.tests.TestCase):
+    def setup_method(self):
+        self.domain = gs.linspace(-math.pi, math.pi)
```
Comment (ninamiolane) [inline tests/tests_geomstats/test_functions.py:15]: @kiranvad I believe replacing this by:
`gs.linspace(-gs.pi, gs.pi, num-50)` would solve the remaining issues.

Comment (kiranvad) [reply tests/tests_geomstats/test_functions.py]: Thanks for the suggestion @ninamiolane 
I fixed this but the tests still fail with a pytorch backend because `gs.vectorize` thrown an error. I suspect this is because I make use of the `signature` parameter that exists in NumPy but not in PyTorch. I am not sure how to fix this, so for now I made the tests only for NumPy and autograd.

File: geomstats/geometry/functions.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+import math
+
+import numpy as np
+
+import geomstats.backend as gs
+from geomstats.geometry.manifold import Manifold
+from geomstats.geometry.riemannian_metric import RiemannianMetric
+
+
+class SinfSpaceMetric(RiemannianMetric):
```
Comment (nguigs) [inline geomstats/geometry/functions.py:12]: Could we use a more explicit class name? Sinf refers to the notation and does not read well if you don't know what it stands for.

File: geomstats/geometry/functions.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    r"""A Riemannian metric on the S_{\inf} space
+
+    Parameters:
+    -------
+    domain_samples : array of shape (n_samples, )
+        grid points on the domain
```
Comment (nguigs) [inline geomstats/geometry/functions.py:18]: Add a capital letter and a full stop.

File: geomstats/geometry/functions.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            Optional, default: None.
+
+        Returns
+        -------
+        inner_product : array-like, shape=[...,]
+            Inner-product.
+        """
+
+        def _f(v1, v2):
+            return gs.cast(np.trapz(v1 * v2, x=self.x), gs.float32)
```
Comment (nguigs) [inline geomstats/geometry/functions.py:49]: The `trapz` method could be added to the backend, see the equivalent in [torch](https://pytorch.org/docs/stable/generated/torch.trapezoid.html) and [tensorflow](https://www.tensorflow.org/probability/api_docs/python/tfp/math/trapz).

File: geomstats/geometry/functions.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+        inner_prod = gs.vectorize(
```
Comment (nguigs) [inline geomstats/geometry/functions.py:51]: There is an `axis` parameter in the `trapz` that should allow to vectorize your computations without using the `gs.vectorize` method. That could help making the code work with other backends.

File: geomstats/geometry/functions.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        -------
+        exp : array-like, shape=[..., n_evals]
+            Point on the hypersphere equal to the Riemannian exponential
+            of tangent_vec at the base point.
+        """
+
+        def _f(v, p):
+            norm_v = self.norm(v)
+            t1 = gs.cos(norm_v) * p
+            t2 = (gs.sin(norm_v) / norm_v) * p
```
Comment (nguigs) [inline geomstats/geometry/functions.py:81]: All these operations are already vectorized or can be by using einsum, please have a look at the finite dim equivalent. Also the use of taylor_exp_even_func make sure we stay precise around 0.
https://github.com/geomstats/geomstats/blob/d89ee0a4eb8cd178a5de5bccc095fda52d9c0732/geomstats/geometry/hypersphere.py#L743-L747

File: geomstats/geometry/functions.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        Returns
+        -------
+        log : array-like, shape=[..., n_evals]
+            Tangent vector at the base point equal to the Riemannian logarithm
+            of point at the base point.
+        """
+
+        def _f(p0, p1):
+            theta = gs.arccos(self.inner_product(p1, p0, base_point=None))
+            return (p1 - p0 * gs.cos(theta)) * (theta / gs.sin(theta))
```
Comment (nguigs) [inline geomstats/geometry/functions.py:113]: Same comment as above.

File: geomstats/geometry/functions.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        ----------
+        point : array-like, shape=[..., dim]
+            discrete evaluation of a function.
+        Returns
+        -------
+        projected_point : array-like, shape=[..., dim]
+            Point projected to the hypersphere.
+        """
+        norm_p = self.metric.norm(point)
+        return point / norm_p
```
Comment (nguigs) [inline geomstats/geometry/functions.py:161]: Does this work for multiple samples?

File: geomstats/geometry/functions.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+        base_point : array-like, shape=[..., dim]
+            Point on the manifold.
+
+        Returns
+        -------
+        tangent_vec : array-like, shape=[..., dim]
+            Tangent vector at base point.
+        """
+        inner_product = self.metric.inner_product(vector, base_point)
+        tangent_vec = vector - inner_product
```
Comment (nguigs) [inline geomstats/geometry/functions.py:220]: Is this right? inner_product is a scalar, shouldn't it be multiplied by a base_point?

File: tests/tests_geomstats/test_functions.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        self.manifold = SinfSpace(self.domain)
+
+    @geomstats.tests.np_and_autograd_only
+    def test_manifold(self):
+        proj_f = self.manifold.projection(self.f)
+        result = self.manifold.belongs(proj_f)
+        self.assertTrue(result, "Expected True but got %s" % result)
+
+    @geomstats.tests.np_and_autograd_only
+    def test_metric(self):
```
Comment (nguigs) [inline tests/tests_geomstats/test_functions.py:24]: Could you add lower level tests (test_inner_product, test_exp, test_log, test_log_and_exp, etc?)

File: tests/tests_geomstats/test_functions.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        num_funcs = 10
+        point_a = self.funcs(1)
+        point_b = self.funcs(2)
```
Comment (nguigs) [inline tests/tests_geomstats/test_functions.py:27]: Could you also test with multiple inputs?

**Review by nguigs — CHANGES_REQUESTED** [review]
Good input, thanks! 
- I made some comments that will allow to generalize the code to other backends.
- The code does not seem vectorized everywhere, have a look at the Hypersphere class to see how to do.
- In fact this manifold and its metric are very close to the Hypersphere ones, could we reuse code from there instead of rewriting it? I guess the only substantial change is the computation of the inner_product.

**Comment by ninamiolane** [comment]
@kiranvad any update on this?

**Comment by kiranvad** [comment]
Hi @ninamiolane and @nguigs 
Thanks for the feedback. Still working through them. Got back from some time away from Seattle recently. Will try to push changes in a couple of weeks or so.

**Comment by ninamiolane** [comment]
@kiranvad excellent, keep us in touch, we're looking forward to this.

File: geomstats/geometry/functions.py (hunk 2 above)
Comment (kiranvad) [reply geomstats/geometry/functions.py:12]: Thanks, I am calling this a `HilbertSphere` in the updated code

File: geomstats/geometry/functions.py (hunk 6 above)
Comment (kiranvad) [reply geomstats/geometry/functions.py:81]: Great, this is really helpful. I tried to use as much code as possible but it was not straightforward to define the new metric using the `HypersphereMetric`. The main challenge is we need the embedding metric to be  `L2Metric` on functions. I did not want to define one separately but if the existing `L2Metric` can be updated (somehow), I think I can write the new metric mostly as another sub-class of Hypersphere metric.

File: geomstats/geometry/functions.py (hunk 8 above)
Comment (kiranvad) [reply geomstats/geometry/functions.py:161]: I think the updated one should work! thanks for pointing this out.

File: geomstats/geometry/functions.py (hunk 9 above)
Comment (kiranvad) [reply geomstats/geometry/functions.py:220]: Yeah, that was wrong. I updated this code to do this properly. Thanks for pointing this out again.

**Comment by kiranvad** [comment]
Update the code for now. Will modify the tests accordingly.

**Comment by kiranvad** [comment]
For some reason, I am not able to push changes to this branch. Suspect the error is the [following](https://github.com/psf/black/issues/2964). Will follow up on this.

**Comment by ninamiolane** [comment]
@kiranvad sounds good, let us know if we can help? feel free to join us on slack (link on geomstats readme) if you have any question.

@lpereira95 : would you know what could be happening with this PR?

**Comment by kiranvad** [comment]
Hello @ninamiolane and @nguigs 
An update on this PR. The latest commits should:
1. Incorporate and address changes requested by @nguigs 
2. All the tests related to `geometry/functions.py` pass but I need to implement a backend function for [_trapz](https://github.com/kiranvad/geomstats/blob/0fef30d0ee4f3254b471eb560fadd0f6c9e18177/geomstats/geometry/functions.py#L13) to get this work with PyTorch and TensorFlow. I couldn't find any resources on how to do this (I tried simply adding it to the folders in `_backend` but that seems to break the code). Any suggestions?

**Comment by ninamiolane** [comment]
Hi @kiranvad , thank you for the update! I see that [pytorch has trapz](https://pytorch.org/docs/stable/generated/torch.trapz.html), [tf also](https://www.tensorflow.org/probability/api_docs/python/tfp/math/trapz) and autograd probably has it. 

Thus, the strategy you mention (putting it in `_backend/pytorch/__init__.py` and likewise for the other backends) should work. What broke the code when you tried?

@lpereira95 any chance you could help @kiranvad with the backends? 🙌  Thank you!

File: tests/tests_geomstats/test_functions.py (hunk 11 above)
Comment (kiranvad) [reply tests/tests_geomstats/test_functions.py]: Done!

File: tests/tests_geomstats/test_functions.py (hunk 10 above)
Comment (kiranvad) [reply tests/tests_geomstats/test_functions.py]: Done!

**Comment by lpereira95** [comment]
Sure, I'll add them to the backend this afternoon, so @kiranvad can assume they're available as `gs.trapz`.

**Comment by kiranvad** [comment]
Thanks, @lpereira95, I modified the code to use `gs.trapz` using #1522 and increased the testing to `pytorch` backend as well.

**Comment by kiranvad** [comment]
Hello @ninamiolane,
I think this PR is ready for another review. The test fails are not related to the contents this PR adds. 
Let me know if you'd like to see more changes or if it's good to go.

**Review by ninamiolane — APPROVED** [review]
Looking good, many thanks!!
